modules/homa-lite/src/lib.rs

```diff
@@ -775,20 +815,35 @@ pub mod module {
 			Ok(())
 		}
 
+		/// Construct XCM message and sent it to the relaychain to withdraw_unbonded Staking
+		/// currency. The staking currency withdrew becomes available to be redeemed. Process
+		/// through redeem requests to fullfill the redeem order.
 		#[transactional]
-		fn process_scheduled_unbond(staking_amount: Balance) -> DispatchResult {
-			let msg = Self::construct_xcm_unreserve_message(T::ParachainAccount::get(), staking_amount);
+		fn process_scheduled_unbond(staking_amount_unbonded: Balance) -> DispatchResult {
+			let msg = Self::construct_xcm_unreserve_message(T::ParachainAccount::get(), staking_amount_unbonded);
 
 			let res = pallet_xcm::Pallet::<T>::send_xcm(Here, Parent, msg);
 			log::debug!("on_idle XCM result: {:?}", res);
 			ensure!(res.is_ok(), Error::<T>::XcmFailed);
 
+			// Update storage with the new available amount
+			AvailableStakingBalance::<T>::mutate(|current| {
+				*current = current.saturating_add(staking_amount_unbonded);
+			});
+
+			Self::deposit_event(Event::<T>::ScheduledUnbondWithdrew(staking_amount_unbonded));
+
 			// Now that there's available staking balance, automatically match existing
 			// redeem_requests.
+			Self::process_redeem_requests_with_available_staking_balance()
+		}
+
+		/// Iterate through all redeem requests, then match them with available_staking_balance.
+		/// This should be called when new available_staking_balance becomes available.
+		#[transactional]
+		fn process_redeem_requests_with_available_staking_balance() -> DispatchResult {
 			let mut new_balances: Vec<(T::AccountId, Balance, Permill)> = vec![];
-			let mut available_staking_balance = Self::available_staking_balance()
-				.checked_add(staking_amount)
-				.ok_or(ArithmeticError::Overflow)?;
+			let mut available_staking_balance = Self::available_staking_balance();
 			for (redeemer, (request_amount, extra_fee)) in RedeemRequests::<T>::iter() {
 				// If all the currencies are minted, return.
```

Comment:
related to Bryan's benchmarking comment, Probably need to bound this to a max number of iterations

Reply:
Updated `adjust_total_staking_currency` to add a limit to the number of requests to match. Also calculate weight from it.